### PR TITLE
CookieBar Component

### LINKED ISSUE #365: [FEATURE] Cookiebar component
Replicate the Cookiebar component for use in React projects.

Design and behaviour reference is [Bootstrap Italia Cookiebar component](https://italia.github.io/bootstrap-italia/docs/componenti/cookiebar/) as usual. In short, Cookiebar is shown every time a `cookies_consent` cookie is not present on the client browser -  it sets that cookie to true when "ok" button is clicked. Current JS code can be seen here:
https://github.com/italia/bootstrap-italia/blob/master/src/js/plugins/cookiebar.js

Working example (built with Bootstrap Italia) here: https://italia.github.io/bootstrap-italia/docs/esempi/cookiebar/

### PR DESCRIPTION
**Pull request by ankingcodes**
Fixes #365

#### PR Checklist

- [x] My branch is up-to-date with the Upstream `master` branch.
- [x] The unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

#### Short description of what this resolves:
This will add a new CookieBar component.

#### Changes proposed in this pull request:

- created `CookieBtn.js` in `src/components`
- added `CookieBtn.js` to `index.js`
- created `CookieBar.stories.js` in `stories/CookieBar/`

### CODE REVIEW

**Comment by ankingcodes** [comment]
@francescozaia done !!

**Comment by dej611** [comment]
Hi @ankingcodes , I think it is better to stick to the Cookiebar reference HTML for styling to prevent diverging issues in the future.

File: src/components/CookieBar/CookieBtn.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import {Button} from '../../index'
+
+const CookieBtn = props => {
+    return (
+        <span>
```
Comment (dej611) [inline src/components/CookieBar/CookieBtn.js]: there's no need for this in the component

File: src/components/CookieBar/CookieBtn.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+            <Button color="link">{props.buttonName}</Button>
```
Comment (dej611) [inline src/components/CookieBar/CookieBtn.js]: It is correct to leverage the `Button` component, but I think you have to stick to the reference HTML.
In particular the `Button` component injects a `btn` class that it needs to be overwritten - see the `cssModules` prop in reactstrap: https://reactstrap.github.io/components/buttons/ .
Also you can leverage the `tag` prop of the `Button` component for reject and accept version as in the reference documentation.

File: stories/CookieBar/CookieBar.stories.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
+import { storiesOf } from '@storybook/react'
+import { withA11y } from '@storybook/addon-a11y'
+import { withInfo } from '@storybook/addon-info'
+
+import { CookieBtn } from '../../src'
+import Esempi from '../Collapse/docs/Innestati.md'
```
Comment (dej611) [inline stories/CookieBar/CookieBar.stories.js]: add a TODO comment here to replace it with the correct one.

File: stories/CookieBar/CookieBar.stories.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+const textStyle = {color: "white", fontSize:16 , paddingTop: "20px", paddingLeft: "15px",  };
+const divStyle = {backgroundColor: "#435A70",bottom: 0, left: 0, borderTopLeftRadius: "3px", borderTopRightRadius: "3px"}
+
+const CookieBarExample = () => (
+   <div style={divStyle}>
```
Comment (dej611) [inline stories/CookieBar/CookieBar.stories.js]: maybe it is worth exploring a componentization of this structure.

**Comment by ankingcodes** [comment]
hello @dej611 , thank you for your reviews. I'll work on them :)

**Comment by ankingcodes** [comment]
@francescozaia I have worked on the changes requested by @dej611 and now the CookieBar has the following features : 
- "Accetto" button closes the CookieBar
- "Preferenzo" button redirects to [Privacy Policy page](https://designers.italia.it/privacy-policy/)
- [Todo] Docs is yet to be written and added to `CookieBar.stories.js`
Kindly review last commit and let me know if I have to write the docs myself or it can be done in another issue !!

File: stories/CookieBar/CookieBar.stories.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import { withA11y } from '@storybook/addon-a11y'
+import { withInfo } from '@storybook/addon-info'
+
+import {Button} from '../../src' 
+//#[TODO] replace docs with correct file
+import Esempi from '../Collapse/docs/Innestati.md'
+
+
+//styling of the CookieBar
+const textStyle = {color: "white",
```
Comment (dej611) [inline stories/CookieBar/CookieBar.stories.js]: Drop all these styling. There's no need to use custom styling, just classes.

File: stories/CookieBar/CookieBar.stories.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      marginLeft:"100px",
+      paddingLeft:"50px",
+      display:"grid",
+      gridTemplateColumns:"auto auto",
+      paddingTop:"14px"
+}
+
+
+const CookieBarExample = (props) => (
+   <div style={divStyle} >
```
Comment (dej611) [inline stories/CookieBar/CookieBar.stories.js]: Convert this to a `CookieBar` component.

File: stories/CookieBar/CookieBar.stories.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+     <p style={textStyle}>Questo sito utilizza cookie tecnici, analytics e di terze parti. <br/> Proseguendo nella navigazione accetti l’utilizzo dei cookie.</p>
+      <div style={divStyle2}>
+          <a href="https://designers.italia.it/privacy-policy/" style={{backgroundColor:"#435A70" ,color:"#E6ECF2",textTransform:"uppercase", fontSize:14, fontWeight:"bold"}}>Preferenzo</a>
+          <Button color="link" style={{backgroundColor:"#435A70",color:"#E6ECF2",textTransform:"uppercase",fontSize:14,
```
Comment (dej611) [inline stories/CookieBar/CookieBar.stories.js]: Drop this styling, no need for this.

File: stories/CookieBar/CookieBar.stories.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+          fontWeight:"bold",marginTop:-24,borderWidth:0}} onClick={props.addTrip}>accetto</Button>
+      </div>
+   </div>
+)
+const Empty = () => (
+  <div>
+  </div>
+)
+
+class App extends Component{
```
Comment (dej611) [inline stories/CookieBar/CookieBar.stories.js]: No need for an app component.
If you need a way to manage the state, you can use the knobs for that.

**Comment by dej611** [comment]
@ankingcodes in my opinion, we need to do a step back and think what it means to create a CookieBar component.

This is the plain component HTML:
```html
<div class="cookiebar">
    <p>Questo sito utilizza cookie tecnici, analytics e di terze parti. <br>Proseguendo nella navigazione accetti l’utilizzo dei cookie.</p>
    <div class="cookiebar-buttons">
        <a href="#" class="cookiebar-btn">Preferenze<span class="sr-only">cookies</span></a>
        <button data-accept="cookiebar" class="cookiebar-btn cookiebar-confirm">Accetto<span class="sr-only"> i cookies</span></button>
    </div>
</div>
```

No custom styling is present in here, just classes.
When moving it to React, usually you try to define what's a component and what's not. So let's try to mockup some componentization before jumping on the code:


```jsx
<CookieBar> //<- the container
  <p>Questo sito utilizza cookie tecnici, analytics e di terze parti. <br>Proseguendo nella navigazione accetti l’utilizzo dei cookie.</p> // <- This can be custom, so not worth to componentize it
  <CookieBarButtons>  // <- this has a specific meaning in a cookie bar, so maybe it's a component
    <CookieBarButton href="#">Preferenze<span class="sr-only">cookies</span></CookieBarButton> <- This has custom classes, so why not make it a cookiebar component out of it?
    <CookieBarButton onClick={...}>Accetto<span class="sr-only"> i cookies</span></CookieBarButton> <- Same here. Pay attention that the tag is different from above
  </CookieBarButtons>
</CookieBar>
```

So, if this make sense, we can move to create the three new components.

**Comment by ankingcodes** [comment]
hey @dej611, thanks for review. I have understood the componentization  part. I tried doing that before. However how will each of the components be styled? Will it be imported from reactstrap?

**Comment by dej611** [comment]
If you want you can recycle the classes and structure there, the `bootstrap italia` css framework and `reactstrap` are available in the toolkit.
For instance in a possible `CookieBar` component you may do:

```jsx
const CookieBar = ({className, ...attributes}) => {
  const classes = /* combine "cookiebar" class with the ones passed by the user here */
  return <div className={classes} {...attributes}/>
}
```

That is a very naïve implementation. For instance you may want the user to be able to change the tag (`div` => something else), or maybe something else.
Note that the `children` prop is included in the `attributes` in here, so no particular code is required to handle that.

**Comment by ankingcodes** [comment]
@dej611 Recycling the classes of reactstrap and bootstrap italia css modules won't be proficient. I'm thinking of customly styling each component. It would be easier for user to change them according to their needs.

**Comment by dej611** [comment]
@ankingcodes recycling the classes from bootstrap italia is the way the toolkit should work. Custom styling is the very last resource, just look at how other components are made in the toolkit.

**Comment by francescozaia** [comment]
> @ankingcodes recycling the classes from bootstrap italia is the way the toolkit should work. Custom styling is the very last resource, just look at how other components are made in the toolkit.

I agree with @dej611: Design React Kit and Bootstrap Italia should use the same classes and semantics to ensure consistency. These libraries are conceived to be solid and rigorous, as they're going to be used for public services, so we're not encouraging users to style or personalise them. In brief, I would stick to Bootstrap Italia HTML code and classes as much as possible.

**Comment by ankingcodes** [comment]
@dej611 @francescozaia got it ! I shall be using `bootstrap-italia/dist/css/bootstrap-italia.min.css` then. Thanks for your guidance.

**Comment by ankingcodes** [comment]
I'm closing this PR. I have made a new PR in reference to this PR.

**Pull request closed without merging**